[Handshake Reassembly] RFC - proposed design doc (#3817)
ARMmbed/mbedtls: docs/proposed/hs_reassembly.md

# Reassembly of incoming fragmented handshake messages

# Summary

> At the moment, TLS handshake fragmentation is not supported. In particular
> this limits the max certificate length that Mbed TLS client can receive.
> 
> This PR adds the ability to reassemble the fragmented handshake messages.

Reviewer: If we want to integrate this design document into the library, we shouldn't reference the PR from which it originates but focus on the feature that's being documented.

```suggestion

This document describes the design of the implementation of reassembly of incoming
TLS handshake messages. The fragmentation of outgoing TLS handshake messages
is currently not supported in Mbed TLS.

Fragmentation and reassemly of DTLS handshake messages is supported in Mbed TLS,
but not documented here.
```

Ability to fragment outgoing handshake messages is out of scope of this PR.

## Goals

1. Full backwards compatibility with the "development" branch of Mbed TLS,
   unless the user explicitly enables fragmentation support in `config.h` and
   via `mbedtls_ssl_conf_xxx()` API.
2. Minimizing the conceptual changes to the Mbed TLS codebase.
3. Robustness and simplicity.

## Non-goals

The below goals are out of scope of this PR, but will be addressed eventually.

> 1. Minimal memory footprint when the fragmentation is enabled.

Reviewer: ```suggestion
1. Minimal memory footprint (RAM and ROM) when the reassembly feature is enabled.
```

> 2. Fragmenting the outgoing messages.

Reviewer: ```suggestion
2. The fragmentation of outgoing TLS handshake messages.
```

# Detailed description of the problem

Mbed TLS can only processs handshake messages that are fully contained within a
single TLS record. If a handshake message spans multiple TLS records, Mbed TLS

> aborts the handshake.

Reviewer: ```suggestion
by default aborts the handshake.
```

> This ability is needed when the remote endpoint sends several concatenated handshake
> messages, e.g. `Server Hello || Server Certificate || Server Key Exchange ||
> Server Done`. If the length of the concatenated sequence exceeds the maximal
> length of a TLS record, the remote endpoint may chose to split a handshake
> message at the TLS record boundary.

Reviewer: ```suggestion
The ability to reassembly fragmented TLS handshake messages is needed when the peer
sends several concatenated handshake messages, e.g. 
```
Server Hello || Server Certificate || Server Key Exchange || Server Done
``` 
If the length of the concatenated sequence exceeds the maximal
length of a TLS record that the peer is willing to support, the peer 
may chose to split a handshake message at the TLS record boundary.
```

## Examples of fragmented handshake messages

The simple scenario assumes that the fragmented handshake message starts and
ends at the record boundary:

```
+---------------------------+    +---------------------+
|TLS                        |    |TLS                  |
|hdr | payload              |    |hdr | payload        |
+----+----------------------+    +----+----------------+
     | +------+------------+|         | +------------+ |
     | |HS Hdr|Body Frag 0 ||  ...    | |Body Frag #n| |
     | +------+------------+|         | +------------+ |
     +----------------------+         +----------------+
```

More general scenario includes two TLS records (`m` and `m+1`), where:

1. Record `m` contains the last fragment of handshake message `k-1`, followed by
   the start of message `k`.
2. Record `m+1` contains the end of message `k`, followed by start of message `k+1`.

```

> +--------------------------------------+  +-------------------------------------+
> |TLS msg #m                            |  |TLS  #m+1                            |
> |hdr | payload                         |  |hdr | payload                        |
> +----+---------------------------------+  +----+--------------------------------+

Reviewer: ```suggestion
+--------------------------------------+  +-------------------------------------+
|TLS record #m                         |  |TLS record #m+1                      |
|hdr | payload                         |  |hdr | payload                        |
+----+---------------------------------+  +----+--------------------------------+
```

     |  +--------+-------------------+ |       | +------------+------+--------+ |
     |  |.. #k-1 |#k ...             | |       | |... #k      | #k+1       ...| |
     |  +--------+-------------------+ |       | +------------+------+--------+ |
     +---------------------------------+       +--------------------------------+
```

> # Proposed solution

Reviewer: As above, I think this should be written under the assumption that the PR makes it into the library.
```suggestion
# Design of TLS handshake reassembly
```

> The suggested approach is using `mbedtls_reader` to accumulate the fragments of
> incoming handshake messages. Contents of TLS records that carry handshake data
> are appended to the accumulator, and the handshake state machine consumes
> messages from the accumulator:

Reviewer: ```suggestion
```suggestion
The library is using the `mbedtls_reader` structure to accumulate the handshake
fragments contained in a sequence of TLS records and presenting them as if they'd
reside in a contiguous buffer. 

Contents of TLS records that carry handshake data are fed to the reader, and the handshake
state machine consumes the handshake data as the granularity of handshake messages. The
necessary accumulation and reassembly is done within the reader.
```

Author: Addressed.

 +-------------------------+ +--------------------------+
                             |
                             +------+
                                    v
                           +----------------+
                           |  Accumulator   |
                           +----------------+
                                    |
                                    +----------+
                                               v
                               +----------+ +-----------------+
                               |   HS k   | |     HS k+1      |
                               +----------+ +-----------------+

```

Conceptually, the `mbedtls_reader` is a FIFO buffer, that offers the following API:

- `mbedtls_reader_feed()`    - appends data to the "tail" of the FIFO.
- `mbedtls_reader_get()`     - provides access to the "head" of the FIFO.
- `mbedtls_reader_commit()`  - informs the "FIFO" that the data that have been previously
                               requested via `mbedtls_reader_get()` have been processed.
- `mbedtls_reader_reclaim()` - cleans up the space that's occupied by the already committed
                               messages.

> ## Data structures

Reviewer: ```suggestion
## API Changes
```

Author: Addressed.

> Reassembly must be enabled by uncommenting `MBEDTLS_SSL_HS_REASSEMBLY`
> in the `include/mbedtls/config.h` header file, and also enabled in the runtime
> by setting a non-zero max message size.

Reviewer: ```suggestion
Support for TLS handshake reassembly is controlled by the compile-time configuration
option `MBEDTLS_SSL_HS_REASSEMBLY`, disabled by default. The maximum amount 
of heap memory used for reassembly is controlled through 
`MBEDTLS_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE` at compile-time and 
through `mbedtls_ssl_conf_hs_reassembly_set_max_message_size()` at runtime.

```

Author: Addressed.

> Suggested configuration preprocessor constructs:

Reviewer: ```suggestion
Example configuration:
```

```
/** Enable handshake reassembly. */
#define MBEDTLS_SSL_HS_REASSEMBLY

/** Hard limit for the maximal message size. */
#define MBEDTLS_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE

/** The operation can not be performed because of the maximal message size limitation. */

> #define MBEDTLS_ERR_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE_EXCEEDED

Reviewer: Suggestion: Add a minimal example of where to call the new runtime-API, and explicitly state if anything else needs to be done by the programmer in order to make use of this feature.

Reviewer: All API-related information should be contained in `ssl.h` and `config.h`, too. The implementation-details can be left to this document.

```

Suggested API to enable reassembly in runtime:

```
struct mbedtls_ssl_config {

...

#if defined(MBEDTLS_SSL_HS_REASSEMBLY)
    /** Maximal supported message size. Value 0 indicates that the reassembly is disabled.*/
    size_t hs_reassembly_max_message_size;
#endif /* MBEDTLS_SSL_HS_REASSEMBLY */

...

};

/**
 * Set max size of reassembled message.
 *
 * \param conf mbedtls config.
 * \param max_message_size - maximal allowed size for the reassembled message.
 *                           Must not exceed MBEDTLS_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE.
 * \return 0 on success, or MBEDTLS_ERR_SSL_HS_REASSEMBLY_MAX_MESSAGE_SIZE_EXCEEDED on failure.
 */
int mbedtls_ssl_conf_hs_reassembly_set_max_message_size( mbedtls_ssl_config *conf, size_t max_message_size );
```

> Suggested implementation in `mbedtls_ssl_context`:

Reviewer: This should be a separate section `## Implementation details` or similar.

Author: Addressed in the design doc.

```
struct mbedtls_ssl_context {

...

#if defined(MBEDTLS_SSL_HS_REASSEMBLY)

>      /* reassembly section */

Reviewer: Can/Should this be moved to `mbedtls_ssl_handshake_params`?

Reviewer: It seems that this has already happened in the actual code :+1:

Author: I will make an additional step over the design document to ensure it is in line with the code.

# Defragmenting handshake messages based on MPS reader

# Summary

At the momement, TLS handshake fragmentation is not supported. In particular
this limits the max certificate length that mbedTLS client can receive.

## Goals

> 1. Full backwards compatibility with the mbedTLS 2.16 version, unless the user

Author: This should read "with development". 2.16 is an atavism, I will update this shortly.

Author: Addressed in 5d5d45f
